common/http-api-client/src/dns.rs

```diff
@@ -0,0 +1,186 @@
+                iter: lookup.into_iter(),
+            });
+            Ok(addrs)
+        })
+    }
+}
+
+impl Iterator for SocketAddrs {
+    type Item = SocketAddr;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|ip_addr| SocketAddr::new(ip_addr, 0))
+    }
+}
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+impl HickoryDnsResolver {
+    pub async fn resolve_str(&self, name: &str) -> Result<LookupIp, BoxError> {
+        let resolver = self.state.get_or_try_init(new_resolver)?;
+
+        // try the primary DNS resolver that we set up (DoH or DoT or whatever)
+        let lookup = match resolver.lookup_ip(name).await {
+            Ok(res) => res,
+            Err(e) => {
+                // on failure use the fall back system configured DNS resolver
+                warn!("primary DNS failed w/ error {e}: using system fallback");
+                let resolver = self.fallback.get_or_try_init(new_resolver_system)?;
+                resolver.lookup_ip(name).await?
+            }
+        };
+
+        Ok(lookup)
+    }
+}
+
+/// Create a new resolver with a custom DoT based configuration. The options are overridden to look
+/// up for both IPv4 and IPv6 addresses to work with "happy eyeballs" algorithm.
+fn new_resolver() -> Result<TokioAsyncResolver, HickoryDnsSystemConfError> {
+    let mut name_servers = NameServerConfigGroup::google_tls();
+    name_servers.merge(NameServerConfigGroup::google_https());
+    // name_servers.merge(NameServerConfigGroup::google_h3());
+    name_servers.merge(NameServerConfigGroup::quad9_tls());
+    name_servers.merge(NameServerConfigGroup::quad9_https());
+    name_servers.merge(NameServerConfigGroup::cloudflare_tls());
+    name_servers.merge(NameServerConfigGroup::cloudflare_https());
+
+    let config = ResolverConfig::from_parts(None, Vec::new(), name_servers);
+
+    let mut opts = ResolverOpts::default();
+    opts.ip_strategy = LookupIpStrategy::Ipv4AndIpv6;
+    // Would like to enable this when 0.25 stabilizes
+    // opts.server_ordering_strategy = ServerOrderingStrategy::RoundRobin;
+
+    Ok(TokioAsyncResolver::tokio(config, opts))
+}
+
+/// Create a new resolver with the default configuration, which reads from the system DNS config
+/// (i.e. `/etc/resolve.conf` in unix). The options are overridden to look up for both IPv4 and IPv6
+/// addresses to work with "happy eyeballs" algorithm.
+fn new_resolver_system() -> Result<TokioAsyncResolver, HickoryDnsSystemConfError> {
+    let (config, mut opts) =
+        hickory_resolver::system_conf::read_system_conf().map_err(HickoryDnsSystemConfError)?;
+    opts.ip_strategy = LookupIpStrategy::Ipv4AndIpv6;
+    Ok(TokioAsyncResolver::tokio(config, opts))
+}
+
+impl fmt::Display for HickoryDnsSystemConfError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("error reading DNS system conf for hickory-dns")
+    }
+}
+
+impl std::error::Error for HickoryDnsSystemConfError {
```

Comment:
Is there a reason why Error is implemented manually instead of just `#[derive(thiserror::Error)]`? Same with `impl fmt::Display`

Reply:
Done, this was implemented manually in the reqwest hickory resolver, so I was just working from their example. I also didn't know we imported thiserror. Easy change :+1: